File: packages/sdk/src/instructions/store-commitment.ts
/**
 * Store Commitment Instruction Builder
 *
 * Stores commitments as Light Protocol compressed accounts
 * Called after transact to store output commitments
 */

import {
  PublicKey,
  ComputeBudgetProgram,
} from '@solana/web3.js';
import { Program } from '@coral-xyz/anchor';
import BN from 'bn.js';

import { derivePoolPda } from './constants';
import { LightProtocol, LightStoreCommitmentParams } from './light-helpers';

/**
 * Store commitment parameters
 */
export interface StoreCommitmentParams {
  /** Token mint */
  tokenMint: PublicKey;
  /** Commitment to store */
  commitment: Uint8Array;
  /** Leaf index in tree */
  leafIndex: bigint;
  /** Stealth ephemeral pubkey (64 bytes) for deriving decryption key */
  stealthEphemeralPubkey: Uint8Array;
  /** Encrypted note data */
  encryptedNote: Buffer;
  /** Relayer public key */
  relayer: PublicKey;
}

/**
 * Build store_commitment transaction using Anchor program
 */
export async function buildStoreCommitmentWithProgram(
  program: Program,
  params: StoreCommitmentParams,
  rpcUrl: string
): Promise<any> {
  const programId = program.programId;
  const lightProtocol = new LightProtocol(rpcUrl, programId);

  // Derive PDAs
  const [poolPda] = derivePoolPda(params.tokenMint, programId);

  // Get Light Protocol validity proof for commitment address
  const commitmentAddress = lightProtocol.deriveCommitmentAddress(poolPda, params.commitment);
  const validityProof = await lightProtocol.getValidityProof([commitmentAddress]);
  const { accounts: remainingAccounts, outputTreeIndex, addressTreeIndex } = lightProtocol.buildRemainingAccounts();

  const convertedProof = LightProtocol.convertCompressedProof(validityProof);
  const addressTreeInfo = {
    addressMerkleTreePubkeyIndex: addressTreeIndex,
    addressQueuePubkeyIndex: addressTreeIndex,
    rootIndex: validityProof.rootIndices[0] ?? 0,
  };

  // Build transaction using Anchor
  // Use Buffer for encryptedNote (Vec<u8>) - same as shield instruction
  const tx = await program.methods
    .storeCommitment({
      commitment: Array.from(params.commitment),
      leafIndex: new BN(params.leafIndex.toString()),
      stealthEphemeralPubkey: Array.from(params.stealthEphemeralPubkey),
      encryptedNote: Buffer.from(params.encryptedNote),  // Buffer, not number[]
      validityProof: convertedProof,
      addressTreeInfo,
      outputTreeIndex,
    })
    .accountsStrict({
      pool: poolPda,
      relayer: params.relayer,
    })
    .remainingAccounts(remainingAccounts)
    .preInstructions([
      ComputeBudgetProgram.setComputeUnitLimit({ units: 600_000 }),
      ComputeBudgetProgram.setComputeUnitPrice({ microLamports: 50000 }),
    ]);

  return tx;
}

/**
 * Build and execute store_commitment transactions for multiple commitments
 *
 * After transact, call this to store each output commitment
 * Note: Anchor's .rpc() already handles confirmation - no extra verification needed
 */
export async function storeCommitments(
  program: Program,
  tokenMint: PublicKey,
  commitments: Array<{
    commitment: Uint8Array;
    leafIndex: bigint;
    stealthEphemeralPubkey: Uint8Array;
    encryptedNote: Buffer;
  }>,
  relayer: PublicKey,
  rpcUrl: string
): Promise<string[]> {
  const signatures: string[] = [];

  for (const commitment of commitments) {
    const tx = await buildStoreCommitmentWithProgram(
      program,
      {
        tokenMint,
        commitment: commitment.commitment,
        leafIndex: commitment.leafIndex,
        stealthEphemeralPubkey: commitment.stealthEphemeralPubkey,
        encryptedNote: commitment.encryptedNote,
        relayer,
      },
      rpcUrl
    );

    const sig = await tx.rpc();
    signatures.push(sig);
  }

  return signatures;
}

File: programs/cloakcraft/src/instructions/pool/store_commitment.rs
//! Store a commitment as a Light Protocol compressed account
//!
//! Called after transact to persist output commitments on-chain.
//! Each commitment requires its own validity proof and transaction.
//!
//! Note: Light Protocol validity proofs with multiple addresses are designed
//! for cross-program operations (e.g., Program A + Program B sharing one proof),
//! NOT for batching multiple creations in a single CPI call.

use anchor_lang::prelude::*;

use crate::state::{Pool, LightValidityProof, LightAddressTreeInfo};
use crate::constants::seeds;
use crate::light_cpi::{create_commitment_account, vec_to_fixed_note};

#[derive(Accounts)]
pub struct StoreCommitment<'info> {
    /// Pool
    #[account(
        seeds = [seeds::POOL, pool.token_mint.as_ref()],
        bump = pool.bump,
    )]
    pub pool: Account<'info, Pool>,

    /// Relayer/submitter (pays for compressed account creation)
    #[account(mut)]
    pub relayer: Signer<'info>,

    // Light Protocol accounts are passed via remaining_accounts
}

/// Parameters for storing a single commitment
#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct StoreCommitmentParams {
    /// The commitment hash
    pub commitment: [u8; 32],
    /// Leaf index (read from commitment counter before transact)
    pub leaf_index: u64,
    /// Stealth ephemeral pubkey for deriving decryption key (64 bytes: X + Y)
    /// If all zeros, decrypt with original spending key (internal operations)
    pub stealth_ephemeral_pubkey: [u8; 64],
    /// Encrypted note data (Vec to avoid heap issues during deserialization)
    pub encrypted_note: Vec<u8>,
    /// Validity proof for the commitment address (non-inclusion)
    pub validity_proof: LightValidityProof,
    /// Address tree info
    pub address_tree_info: LightAddressTreeInfo,
    /// Output state tree index
    pub output_tree_index: u8,
}

pub fn store_commitment<'info>(
    ctx: Context<'_, '_, '_, 'info, StoreCommitment<'info>>,
    params: StoreCommitmentParams,
) -> Result<()> {
    let pool = &ctx.accounts.pool;

    // Convert Vec to fixed array (avoids heap issues during deserialization)
    let (encrypted_note_arr, encrypted_note_len) = vec_to_fixed_note(&params.encrypted_note);

    // Create commitment compressed account with full encrypted note inline
    // This allows scanning via Light Protocol API without external indexer
    create_commitment_account(
        &ctx.accounts.relayer.to_account_info(),
        ctx.remaining_accounts,
        params.validity_proof,
        params.address_tree_info,
        params.output_tree_index,
        pool.key(),
        params.commitment,
        params.leaf_index,
        params.stealth_ephemeral_pubkey,
        encrypted_note_arr,
        encrypted_note_len,
    )?;

    msg!("Commitment stored: leaf_index={}", params.leaf_index);

    Ok(())
}
